### ISSUE TITLE
Unmodified files with DOS line endings show up as modified 

### ISSUE THREAD

**Issue by rahulsalvi** [lewis6991/gitsigns.nvim#369]
### Description

When I open a file with DOS line endings (CRLF), the entire buffer shows up as modified, even though the file is unchanged. If I tell nvim to view the file with unix line endings (LF), there is a ^M character at the end of each line, but the buffer does not show up as modified. When I git bisect the issue, the first bad commit is 82ac003.

### Neovim version

`
NVIM v0.5.0 Build type: Release LuaJIT 2.0.5 Compilation: /usr/bin/cc -D_FORTIFY_SOURCE=2 -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -DNVIM_TS_HAS_SET_MATCH_LIMIT -O2 -DNDEBUG -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-prototypes -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fno-common -fdiagnostics-color=always -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -DMIN_LOG_LEVEL=3 -I/build/neovim/src/neovim-0.5.0/build/config -I/build/neovim/src/neovim-0.5.0/src -I/usr/include -I/build/neovim/src/neovim-0.5.0/build/src/nvim/auto -I/build/neovim/src/neovim-0.5.0/build/include Compiled by builduser  Features: +acl +iconv +tui See ":help feature-compile"     system vimrc file: "$VIM/sysinit.vim"   fall-back for $VIM: "/usr/share/nvim"  Run :checkhealth for more info
`

### Operating system and version

Linux 5.14.3

### Steps to reproduce

1. Commit a file with dos line endings (CRLF) to a git repo (e.g. test.txt)
2. `nvim -nu minimal.lua test.txt` (make sure ff=dos)
3. The entire buffer shows up as modified according to gitsigns.nvim
4. run ":e ++ff=unix" in nvim to open the file in unix mode (ff=unix)
5. The buffer will now not show up as modified according to gitsigns, but you will see the ^M character at the end of each line.

### Expected behavior

There should not be any signs when viewing an unmodified file with DOS line endings.

### Actual behavior

The entire buffer shows up as modified when there are DOS line endings.

### Minimal config

```Lua
vim.cmd [[set runtimepath=$VIMRUNTIME]]
vim.cmd [[set packpath=/tmp/nvim/site]]
local package_root = '/tmp/nvim/site/pack'
local install_path = package_root .. '/packer/start/packer.nvim'

if vim.fn.isdirectory(install_path) == 0 then
  print("Installing Gitsigns and dependencies.")
  vim.fn.system { 'git', 'clone', '--depth=1', 'https://github.com/wbthomason/packer.nvim', install_path }
end

-- Load plugins
require('packer').startup {
  {
    'wbthomason/packer.nvim',
    {
      'lewis6991/gitsigns.nvim',
      requires = { 'nvim-lua/plenary.nvim' },
    },
    -- ADD PLUGINS THAT ARE _NECESSARY_ FOR REPRODUCING THE ISSUE
  },
  config = {
    package_root = package_root,
    compile_path = install_path .. '/plugin/packer_compiled.lua',
    display = { non_interactive = true },
  },
}

_G.load_config = function()
  require('gitsigns').setup{
    debug_mode = true, -- You must add this to enable debug messages
    -- GITSIGNS CONFIG THAT ARE _NECESSARY_ FOR REPRODUCING THE ISSUE
  }

  -- ADD INIT.LUA SETTINGS THAT ARE _NECESSARY_ FOR REPRODUCING THE ISSUE

  -- ALSO INCLUDE GITSIGNS DEBUG MESSAGES
  vim.cmd('Gitsigns debug_messages')
end

require('packer').sync()

vim.cmd [[autocmd User PackerComplete ++once echo "Ready!" | lua load_config()]]
```

**Comment by clason** [comment]
That's the same issue we talked about in chat a short while ago; you can repro this easily by checking out the Neovim repo and looking at `scripts/vim-patch.sh`, lines 186ff.

It's an issue with `nvim_buf_get_lines` silently replacing these with `\r`; `gitsigns` would have to match this mangling before computing the diff.

**Comment by lewis6991** [comment]
https://github.com/lewis6991/gitsigns.nvim/commit/82ac00301a1240c25ba3cb5845c8a62e0ac40f08 probably make the issue worse since it completely preserves `^M` from `git show` whereas plenary jobs did newline processing and probably inadvertently stripped them.

Interestingly opening `scripts/vim-patch.sh` no longer shows an edited line for me on nvim `v0.6.0-dev+291-gd56002f7b`. So not sure what's happened.

@rahulsalvi do you know how to easily generate a file with dos line endings?

**Comment by rahulsalvi** [comment]
`unix2dos` will change line endings in a file from LF to CRLF. `dos2unix` works the other way around. `set ff=dos` on my machine tries to actually modify the buffer by changing all the line endings. Using `:e ++ff=dos` reopens the buffer without changing it but with the dos file format. I haven't tried the latest nvim, only v0.5.0.

**Comment by lewis6991** [comment]
Worked on this a bit and found a case where `\r` wasn't being handled.

Can you try #375 and let me know if it fixes the issue for you?

**Comment by rahulsalvi** [comment]
It looks like this works correctly on the main buffer, but the preview_hunk window still has ^M characters in it. Thanks for the help so far.

**Comment by lewis6991** [comment]
So dealing with the `\r` turns out to be quite a difficult problem.

I've found that just treating `\r` as if it just a normal character (as opposed to treating `\r\n` as a line boundary) seems to simplify things a lot an guarantees diffs will always be accurate. The only problem with that is that we need to carefully strip them out when rendering them.

I've fixed the PR to remove the `\r` when rendering them in `preview_hunk` and `diffthis` ONLY IF the buffer is a dos file AND all lines in the hunk end with `\r`, otherwise they will be rendered. However there are still other actions that will keep them since it's hard to know whether they should be rendered or not. Feel free to raise another issue if you would like them fixed.
